pandas/tests/io/excel/test_readers.py

```diff
@@ -1020,10 +1031,10 @@ def test_excel_read_buffer(self, engine, read_ext):
         tm.assert_frame_equal(expected, actual)
 
     def test_reader_closes_file(self, engine, read_ext):
-        f = open("test1" + read_ext, "rb")
-        with pd.ExcelFile(f) as xlsx:
-            # parses okay
-            pd.read_excel(xlsx, "Sheet1", index_col=0, engine=engine)
+        with open("test1" + read_ext, "rb") as f:
```

Comment:
Can you revert this? Not totally equivalent

Reply:
It is indeed not equivalent, but that is on purpose. The test is faulty. First you open a file and fail to close it, then you check for a file leak. This demonstrates in fact the issue we are having. Ownership of the file handle should be managed at the level it is created, unless you explicitly transfer ownership of the file handle to `ExcelFile`, but IMHO that is not what you would expect (and the reason for the ticket)

Reply:
@WillAyd do you know what the original intent of this test was? The blame history just points to a test reorganization.

Seems like the disagreement is around whether `ExcelFile(obj).__exit__` should close the file when given an open file object.

@roberthdevries can you check the behavior here on pandas 0.25.3 and 1.0.1? And the behavior on this branch is to not close the file when provided an open file?

I don't think this changed test is adding much value. It's just testing the behavior of `open` now, which we can probably rely on being correct :)

Reply:
Yea I think this test strays a little bit from the OP, where a straight ExcelFile object is closing the handle. In this case the context manager is closing it, which maybe is OK. I can't speak to original history of this.

So @roberthdevries can you maybe summarize what you are proposing? Something like:

 1. ExcelFile Context Manager + path: should manage open/close lifecycle
 2. ExcelFile Context Manager + handle: should not manage open/close lifecycle
 3. ExcelFile + path: should manage open, not close
 4. ExcelFile + handle: should not manage open/close lifecycle

? 

I think point 2 above is a little ambiguous; if there's a precedent for it elsewhere would be interested

Reply:
I think the basic philosophy should be that a resource should be released at the same level where it was created. This is one of the main use cases of context managers.
So whenever a resource is released at **another** level, you violate this rule. (hence this ticket)
The original `ExcelFile` code closed a file it got from the outside, and this specific test relied on this (erroneous IMHO) behavior.
So a nested context manager should work as expected, in the sense that the outside context manager releases the outer resource and the inner context manager releases the inner resource.

So:
```python
with open("foo.xlsx") as f:
    with ExcelFile(f) as xlsx:
        # do stuff
```
In this specific example the `ExcelFile` context manager should **not** close `f`, as that would be the responsibility of the `open` context manager.

In this example:
```python
f = open("foo.xlsx")
with ExcelFile(f) as xlsx:
    # do stuff
```
We have a file leak caused by the programmer forgetting to close the file handle. This is a programmer error, and IMHO not something that needs to be fixed.
However, whenever `f` goes out of scope, the `__del__` method will be called which will probably close the file anyway.

For use cases where `ExcelFile` is used in combination with path names the context manager case implies closing the file.
In this case:
```python
xlsx = ExcelFile(filename)
```
I would propose to have a `__del__` method which will close the file and the other `ExcelFile` resources.

So I agree with you on case 1 and 4.
I also agree with case 2, there is no ambiguity for me. The `ExcelFile` context manager only should manage its own resources, not the file resource.
For case 3 there should be a nuance in the sense that the `__del__` method should close the resource.
For case 4, the `ExcelFile` should not touch the open/close lifecyle of the file handle, but the `ExcelFile` resource itself should be closed again in the `__del__` method.

Reply:
Thanks for the insights. So agree with what you are saying but not sure that is clearly expressed in the implementation. Shouldn't we alternately just within the ExcelFile construct or `__enter__` keep track of whether or not the resource lifecycle is owned or borrowed? As is I don't think that's ever very clear

Reply:
> Shouldn't we alternately just within the ExcelFile constructor

That seems most sensible to me. Do we have a standard method for checking if something is a file-like object vs. a path?

Reply:
It ends up not being an issue here since ExcelFile doesn't open any files, just the engine.